### PR TITLE
Sediment model fixes + simpler model

### PR DESCRIPTION
**Pull request by jagoosw**
I've finally fixed sediment models so they work properly and conserve nitrogen!

Also fixed a few bugs in the sediment model time-stepping and adds a simple instant remineralization model (as described in PISCESv2 paper).

Hopefully after #120 is merged the odd commit history will be fixed.

### CODE REVIEW

File: src/Boundaries/Sediments/Sediments.jl [hunk 1]
```diff
@@ -27,13 +26,14 @@ sediment_fields(::AbstractSediment) = ()
 function update_tendencies!(bgc, sediment::FlatSediment, model)
     arch = model.grid.architecture
 
-    for (i, tracer) in enumerate(sediment_tracers(sediment))    
-        launch!(arch, model.grid, :xy, store_flat_tendencies!, sediment.tendencies.Gⁿ[i], sediment.tendencies.G⁻[i])
+    for (i, tracer) in enumerate(sediment_tracers(sediment))
+        launch!(arch, model.grid, :xy, store_flat_tendencies!, sediment.tendencies.G⁻[i], sediment.tendencies.Gⁿ[i])
```
Comment (jagoosw) [inline src/Boundaries/Sediments/Sediments.jl:30]: This was the main issue with conservation because it just wasn't being integrated correctly

File: src/Models/AdvectedPopulations/NPZD.jl [hunk 2]
```diff
@@ -337,4 +340,9 @@ adapt_structure(to, npzd::NPZD) =
 
                                              adapt(to, npzd.particles))
 
+@inline nitrogen_flux(grid, advection, bgc::NPZD, tracers, i, j) = - (div_Uc(i, j, 0, grid, advection, biogeochemical_drift_velocity(bgc, Val(:D)), tracers.D) +
```
Comment (jagoosw) [inline src/Models/AdvectedPopulations/NPZD.jl:343]: I'm not convinced that this is generally true. I'm the column case it is since there is no horizontal velocity, but if the bottom cell didn't have a no slip boundary (or worse had some imposed flux which should give the cell below the boundary a value) then I think this will be wrong. Instead I think I need to borrow some code from a boundary condition to calculate the flux through the boundary properly.

File: test/test_sediment_particles.jl
Comment (jagoosw) [inline test/test_sediment_particles.jl:1]: I guess given that both particles and sediment now work the same way whether or not the other is there then we don't really need this test

**Comment by jagoosw** [comment]
Perhaps I should make sediments work with bathymetry here too

File: src/Boundaries/Sediments/coupled_timesteppers.jl [hunk 3]
```diff
@@ -44,6 +46,25 @@ import Oceananigans.TimeSteppers: ab2_step!, rk3_substep!
     return nothing
 end
 
+@inline function ab2_step!(model::HydrostaticFreeSurfaceModel{<:Any, <:Any, <:AbstractArchitecture, <:Any, <:Any, <:Any, <:Any, <:Any, <:Any, <:Any, <:Any, <:ContinuousFormBiogeochemistry{<:Any, <:FlatSediment}}, Δt, χ)
+    # Step locally velocity and tracers
+    @apply_regionally local_ab2_step!(model, Δt, χ)
+
+    # blocking step for implicit free surface, non blocking for explicit
```
Comment (navidcy) [inline src/Boundaries/Sediments/coupled_timesteppers.jl:53]: What does this mean?

File: src/Boundaries/Sediments/coupled_timesteppers.jl [hunk 4]
```diff
@@ -83,15 +104,14 @@ function rk3_substep!(model::NonhydrostaticModel{<:Any, <:Any, <:Any, <:Any, <:A
         launch!(arch, model.grid, :xy, rk3_step_flat_field!, 
                 field, Δt, γⁿ, ζⁿ,
                 sediment.tendencies.Gⁿ[i], 
-                sediment.tendencies.G⁻[i])
+                sediment.tendencies.G⁻[i])
```
Comment (navidcy) [inline src/Boundaries/Sediments/coupled_timesteppers.jl:107]:
```suggestion
                sediment.tendencies.G⁻[i])
```

File: src/Boundaries/Sediments/simple_multi_G.jl [hunk 5]
```diff
@@ -61,43 +61,44 @@ using OceanBioME, Oceananigans, OceanBioME.Sediments
 
 grid = RectilinearGrid(size=(3, 3, 30), extent=(10, 10, 200))
 
-sediment_model = SimpleMultiG(grid)
+sediment_model = SimpleMultiG(; grid)
```
Comment (navidcy) [inline src/Boundaries/Sediments/simple_multi_G.jl:64]: If this is a doctest for example it’s ensured to be up to date always.

File: src/Boundaries/Sediments/simple_multi_G.jl [hunk 6]
```diff
@@ -138,30 +139,37 @@ adapt_structure(to, sediment::SimpleMultiG) =
...
-        sediment.tendencies.Gⁿ.C_ref[i, j, 1] = max(0.0, sediment.refactory_fraction * carbon_deposition)
+        sediment.tendencies.Gⁿ.C_slow[i, j, 1] = (1 - sediment.refactory_fraction) * sediment.slow_fraction * carbon_deposition - C_min_slow
+        sediment.tendencies.Gⁿ.C_fast[i, j, 1] = (1 - sediment.refactory_fraction) * sediment.fast_fraction * carbon_deposition - C_min_fast
+        sediment.tendencies.Gⁿ.C_ref[i, j, 1] = sediment.refactory_fraction * carbon_deposition
 
-        sediment.tendencies.Gⁿ.N_slow[i, j, 1] = (1 - sediment.refactory_fraction) * sediment.slow_fraction * nitrogen_deposition - sediment.slow_decay_rate * sediment.fields.N_slow[i, j, 1]
-        sediment.tendencies.Gⁿ.N_fast[i, j, 1] = (1 - sediment.refactory_fraction) * sediment.fast_fraction * nitrogen_deposition - sediment.slow_decay_rate * sediment.fields.N_fast[i, j, 1]
-        sediment.tendencies.Gⁿ.N_ref[i, j, 1] = max(0.0, sediment.refactory_fraction * nitrogen_deposition)
+        sediment.tendencies.Gⁿ.N_slow[i, j, 1] = (1 - sediment.refactory_fraction) * sediment.slow_fraction * nitrogen_deposition - N_min_slow
```
Comment (navidcy) [inline src/Boundaries/Sediments/simple_multi_G.jl:170]: inbounds?

Comment (navidcy) [reply src/Boundaries/Sediments/simple_multi_G.jl:170]: Also to the lines below

File: src/Models/AdvectedPopulations/NPZD.jl [hunk 7]
```diff
@@ -37,6 +38,8 @@ import Oceananigans.Biogeochemistry: required_biogeochemical_tracers,
 
 import OceanBioME: maximum_sinking_velocity
 
+import OceanBioME.Boundaries.Sediments: nitrogen_flux, carbon_flux, remineralizaiton_reciever
```
Comment (navidcy) [inline src/Models/AdvectedPopulations/NPZD.jl:41]: Typos (two). Perhaps they propagate further?

File: src/Boundaries/Sediments/coupled_timesteppers.jl (hunk 3 above)
Comment (jagoosw) [reply src/Boundaries/Sediments/coupled_timesteppers.jl:53]: No idea, this is copied from the Oceananigans repo.

I don't think this is the ideal solution but the only was I can currently think of to integrate the sediment fields.

File: src/Boundaries/Sediments/simple_multi_G.jl (hunk 6 above)
Comment (jagoosw) [reply src/Boundaries/Sediments/simple_multi_G.jl:170]: I think this whole block is inboundsed right?

File: src/Boundaries/Sediments/coupled_timesteppers.jl (hunk 3 above)
Comment (navidcy) [reply src/Boundaries/Sediments/coupled_timesteppers.jl:53]: Hm. My question was about the comment. Perhaps is a leftover comment from KernelAbstractions v0.8? Is this comment still in Oceananigans?

File: src/Boundaries/Sediments/simple_multi_G.jl (hunk 6 above)
Comment (navidcy) [reply src/Boundaries/Sediments/simple_multi_G.jl:170]: Yes! I missed it!

File: src/Boundaries/Sediments/coupled_timesteppers.jl (hunk 3 above)
Comment (jagoosw) [reply src/Boundaries/Sediments/coupled_timesteppers.jl:53]: Yeah:
https://github.com/CliMA/Oceananigans.jl/blob/56fc69bcb08623424afe25e39ae616c29208095e/src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_ab2_step.jl#L11C1-L20C4

Comment (navidcy) [reply src/Boundaries/Sediments/coupled_timesteppers.jl:53]: Ok!

File: src/Boundaries/Sediments/simple_multi_G.jl (hunk 5 above)
Comment (navidcy) [reply src/Boundaries/Sediments/simple_multi_G.jl]: I made it a doctest

File: src/Models/AdvectedPopulations/NPZD.jl (hunk 2 above)
Comment (navidcy) [reply src/Models/AdvectedPopulations/NPZD.jl:343]: I'm a bit hesitant with indexing `k=0` here. This is a halo region... What's the thinking here?

Comment (jagoosw) [reply src/Models/AdvectedPopulations/NPZD.jl:343]: For tracers, the first halo cell is filled with the `k=1` value so `div_Uc`:
https://github.com/CliMA/Oceananigans.jl/blob/56fc69bcb08623424afe25e39ae616c29208095e/src/Advection/tracer_advection_operators.jl#L31C12-L33C103
becomes 
`1/Vᶜᶜᶜ(i, j, k, grid) * δzᵃᵃᶜ(i, j, k, grid, _advective_tracer_flux_z, advection, U.w, c) = ` -wC(k=0)Aᵃᵃᶜ/V. Whereas for the `k=1` it is w(C(k=1) - C(k=2))Aᵃᵃᶜ/V

The conservation tests don't pass with `k=1` either.

Comment (jagoosw) [reply src/Models/AdvectedPopulations/NPZD.jl:343]: I guess a more correct way to put this is:
https://github.com/CliMA/Oceananigans.jl/blob/56fc69bcb08623424afe25e39ae616c29208095e/src/Advection/centered_advective_fluxes.jl#L31C1-L31C1
?

Comment (jagoosw) [reply src/Models/AdvectedPopulations/NPZD.jl:343]: Written a better version of this now

File: src/Boundaries/Sediments/instant_remineralization.jl [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        flux = nitrogen_flux(grid, advection, bgc, tracers, i, j) * Δz
+
+        burial_efficiency = sediment.burial_efficiency_constant1 + sediment.burial_efficiency_constant2 * ((flux / 6.56) / (7 + flux / 6.56)) ^ 2
+
+        # sediment evolution
+        sediment.tendencies.Gⁿ.N_storage[i, j, 1] = burial_efficiency * flux
+
+        remineralisation_reciever(bgc, timestepper.Gⁿ)[i, j, 1] += flux * (1 - burial_efficiency) / Δz
+    end
+end
```
Comment (navidcy) [inline src/Boundaries/Sediments/instant_remineralization.jl]:
```suggestion
    Δz = zspacing(i, j, 1, grid, Center(), Center(), Center())

    flux = nitrogen_flux(grid, advection, bgc, tracers, i, j) * Δz

    burial_efficiency = sediment.burial_efficiency_constant1 + sediment.burial_efficiency_constant2 * ((flux / 6.56) / (7 + flux / 6.56)) ^ 2

    # sediment evolution
    @inbounds sediment.tendencies.Gⁿ.N_storage[i, j, 1] = burial_efficiency * flux

    @inbounds remineralisation_receiver(bgc, timestepper.Gⁿ)[i, j, 1] += flux * (1 - burial_efficiency) / Δz
end

```

File: src/Models/AdvectedPopulations/NPZD.jl [hunk 9]
```diff
@@ -337,4 +340,9 @@ adapt_structure(to, npzd::NPZD) =
 
                                              adapt(to, npzd.particles))
 
+@inline nitrogen_flux(grid, advection, bgc::NPZD, tracers, i, j) = sinking_flux(i, j, grid, advection, Val(:D), bgc, tracers) +
+                                                                   sinking_flux(i, j, grid, advection, Val(:P), bgc, tracers)
+                 
+@inline carbon_flux(bgc::NPZD, tracers, i, j) = nitrogen_flux(bgc, tracers, i, j) * 6.56
+@inline remineralisation_receiver(::NPZD, tendenceies) = tendenceies.N
```
Comment (navidcy) [inline src/Models/AdvectedPopulations/NPZD.jl:347]: tendenceies: typo?

**Review by navidcy — APPROVED** [review]

File: src/Boundaries/Sediments/instant_remineralization.jl [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+    flux = nitrogen_flux(grid, advection, bgc, tracers, i, j) * Δz
+
+    burial_efficiency = sediment.burial_efficiency_constant1 + sediment.burial_efficiency_constant2 * ((flux / 6.56) / (7 + flux / 6.56)) ^ 2
+
+    # sediment evolution
+    @inbounds sediment.tendencies.Gⁿ.N_storage[i, j, 1] = burial_efficiency * flux
+
+    @inbounds timestepper.Gⁿ[remineralisation_reciever(bgc)][i, j, 1] += flux * (1 - burial_efficiency) / Δz
+    end
```
Comment (navidcy) [inline src/Boundaries/Sediments/instant_remineralization.jl:95]:
```suggestion
```
